addons/website_sale/controllers/mail.py

```diff
@@ -0,0 +1,23 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from werkzeug.urls import url_encode
+
+from odoo.http import request
+from odoo.addons.mail.controllers import mail
+
+class MailController(mail.MailController):
+
+    @classmethod
+    def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
+        # no model / res_id, meaning no possible record -> direct skip to super
+        if not model or not res_id or model not in request.env:
+            return super()._redirect_to_record(model, res_id, access_token=access_token, **kwargs)
+
+        if (
+            isinstance(request.env[model], request.env.registry["product.template"]) and
+            request.env.user.is_public
+        ):
+            url_params = { "highlight_message_id": kwargs.get("highlight_message_id") }
+            url = f"/shop/{res_id}?{url_encode(url_params)}"
+            return request.redirect(url)
+        return super()._redirect_to_record(model, res_id, access_token=access_token, **kwargs)
```

Comment:
It's strange to have 2 different super calls.
Maybe just have one big if with all the necessary conditions for the override, and call super only once at the end.